Write the getting started with reactive documentation (#7539)
quarkusio/quarkus: docs/src/main/asciidoc/getting-started-reactive.adoc

        .onItem().apply(n -> String.format("hello %s - %d", name, n))
        .transform().byTakingFirstItems(count);
}
----

It generates a greeting message every second and stops after `count` messages.

In the `ReactiveGreetingResource` add the following method:

[source, java]
----
@GET
@Produces(MediaType.APPLICATION_JSON)
@Path("/greeting/{count}/{name}")
public Multi<String> greetings(@PathParam int count, @PathParam String name) {
  return service.greetings(count, name);
}
----

This endpoint streams the items to the client as a JSON Array.
The name and number of messages are parameterized using path parameters.

So calling the endpoint produces something like:
[source, shell]
----
$ curl http://localhost:8080/hello/greeting/3/neo
["hello neo - 0","hello neo - 1","hello neo - 2"]
----

We can also generate Server-Sent Event responses by returning a `Multi`:

[source, java]
----
@GET
@Produces(MediaType.SERVER_SENT_EVENTS)
@SseElementType(MediaType.TEXT_PLAIN)
@Path("/stream/{count}/{name}")
public Multi<String> greetingsAsStream(@PathParam int count, @PathParam String name) {
    return service.greetings(count, name);
}
----

> The only difference with the previous snippet is the produced type and the `@SseElementType` annotation indicating the type of each event.

Reviewer: Maybe add a note on why we need `SseElementType` : 
> the `@SseElementType` annotation indicating the type of each event, as the `@Produces` annotation defines `SERVER_SENT_EVENTS` JAX-RS needs it to knows the content-type of each event.

[source, shell]
----
$ curl -N http://localhost:8080/hello/stream/5/neo
data: hello neo - 0

data: hello neo - 1

data: hello neo - 2

data: hello neo - 3

data: hello neo - 4
----

== Using Reactive APIs

=== Using Quarkus reactive APIs

Quarkus provides many reactive APIs using the Mutiny model.
In this section, we are going to see how you can use the Reactive PostgreSQL driver to interact with your database in a non-blocking and reactive way.

Create a new project using:

[source,shell,subs=attributes+]
----
mvn io.quarkus:quarkus-maven-plugin:{quarkus-version}:create \
    -DprojectGroupId=org.acme \
    -DprojectArtifactId=getting-started-reactive-crud \
    -DclassName="org.acme.reactive.crud.FruitResource" \
    -Dpath="/fruits" \
    -Dextensions="resteasy-mutiny, resteasy-jsonb, reactive-pg-client"
cd getting-started-reactive-crud
----

This application is interacting with a PostgreSQL database, so you need one:

[source, shell]
----
docker run --ulimit memlock=-1:-1 -it --rm=true --memory-swappiness=0 \
           --name postgres-quarkus-reactive -e POSTGRES_USER=quarkus_test \
           -e POSTGRES_PASSWORD=quarkus_test -e POSTGRES_DB=quarkus_test \

>            -p 5432:5432 postgres:10.5

Reviewer: Maybe use a more recent version, 11.2 ?

Author: Done.

The Mutiny version of the Vert.x APIs also offers:

* `andAwait` methods such as `sendAndAwait`. `andAwait` indicates that the caller thread is blocked until the result is available.
Be aware not to block the event loop / IO thread that way.
* `andForget` methods such as `writeAndForget`. `andForget` is available for method returning a `Uni`.
`andForget` indicates that you don't need the resulting `Uni` indicating the success or failure of the operation.
However, remember that if you don't subscribe, the operation would not be triggered.
`andForget` manages this for you and manage the subscription.
* `toMulti` methods allowing to transform a Vert.x `ReadStream` into a `Multi`
* `toBlockingIterable` / `toBlockingStream` methods allowing to transform a Vert.x `ReadStream` into a blocking iterable or blocking `java.util.Stream`

=== Using RxJava or Reactor APIs

Mutiny provides utilities to convert RxJava 2 and Project Reactor types to `Uni` and `Multi`.

RxJava 2 converters are available in the following dependency:

[source, xml]
----
<dependency>
    <groupId>io.smallrye.reactive</groupId>
    <artifactId>mutiny-rxjava</artifactId>

>     <version>$VERSION</version>

Reviewer: version should comes from Quarkus bom or we risk having non-coherent versions of mutiny

Reviewer 2: Yeah, if it's in the BOM, we could skip the version.

Author: It's not in the BOM :-) But I can add them.

Author: Defined in the BOM, documentation updated.

Single<Optional<String>> single = uni.convert().with(UniRxConverters.toSingle());
Single<String> single2 = uni.convert().with(UniRxConverters.toSingle().failOnNull());
Maybe<String> maybe = uni.convert().with(UniRxConverters.toMaybe());
Observable<String> observable = uni.convert().with(UniRxConverters.toObservable());
Flowable<String> flowable = uni.convert().with(UniRxConverters.toFlowable());
// ...
Completable completable = multi.convert().with(MultiRxConverters.toCompletable());
Single<Optional<String>> single = multi.convert().with(MultiRxConverters.toSingle());
Single<String> single2 = multi.convert().with(MultiRxConverters
        .toSingle().onEmptyThrow(() -> new Exception("D'oh!")));
Maybe<String> maybe = multi.convert().with(MultiRxConverters.toMaybe());
Observable<String> observable = multi.convert().with(MultiRxConverters.toObservable());
Flowable<String> flowable = multi.convert().with(MultiRxConverters.toFlowable());
----

Project Reactor converters are available in the following dependency:

[source, xml]
----
<dependency>
    <groupId>io.smallrye.reactive</groupId>
    <artifactId>mutiny-reactor</artifactId>

>     <version>$VERSION</version>

Reviewer: same here : in the bom

</dependency>
----

So if you have an API returning Reactor types (`Mono`, `Flux`), you can create `Unis` and `Multis` as follows:

[source, java]
----
import io.smallrye.mutiny.converters.multi.MultiReactorConverters;
import io.smallrye.mutiny.converters.uni.UniReactorConverters;
// ...
Uni<String> uniFromMono = Uni.createFrom().converter(UniReactorConverters.fromMono(), mono);
Uni<String> uniFromFlux = Uni.createFrom().converter(UniReactorConverters.fromFlux(), flux);

Multi<String> multiFromMono = Multi.createFrom().converter(MultiReactorConverters.fromMono(), mono);
Multi<String> multiFromFlux = Multi.createFrom().converter(MultiReactorConverters.fromFlux(), flux);
----

You can also transform `Unis` and `Multis` into Reactor types:

[source, java]
----
Mono<String> mono = uni.convert().with(UniReactorConverters.toMono());
Flux<String> flux = uni.convert().with(UniReactorConverters.toFlux());

Mono<String> mono2 = multi.convert().with(MultiReactorConverters.toMono());
Flux<String> flux2 = multi.convert().with(MultiReactorConverters.toFlux());
----

> === Using API using CompletionStages or Publisher

Reviewer: Maybe rename the title as 
```
=== Using CompletionStages or Publisher API
```
Th current one feels strange to me.

Author: Fixed.

////
This guide is maintained in the main Quarkus repository
and pull requests should be submitted there:
https://github.com/quarkusio/quarkus/tree/master/docs/src/main/asciidoc
////
= Quarkus - Getting started with Reactive

include::./attributes.adoc[]

:toc: macro
:toclevels: 4
:doctype: book

> :icons: font

Reviewer 2: Please don't use that here, we want this config to be global and not per guide.

Author: I copied it from another file

Author: (the original getting-started)

Reviewer 2: OK, I suppose it will work fine then.

Author: I've removed them, wondering if we should remove it from the original getting-started.

Reviewer 2: I think it would be safer. If one day we want to change it, we will want to do it globally.

> :docinfo1:

Reviewer 2: That one looks suspicious too.

image:http-blocking-sequence.png[alt=Behavior when using the imperative routes, 50%]

But if the HTTP request targets a reactive (non-blocking) route, the routing layer invokes the route on the IO thread giving lots of benefits such as higher concurrency and performance:

image:http-reactive-sequence.png[alt=Behavior when using the reactive routes, 50%]

As a consequence, many Quarkus components are designed with reactive in mind, such as database access (PostgreSQL, MySQL, Mongo, etc.), application services (mail, template engine, etc.), messaging (Kafka, AMQP, etc.) and so on.
But, to fully benefit from this model, the application code should be written in a non-blocking manner.
That’s where having a reactive API is an ultimate weapon.

== Mutiny - A reactive programming library

https://github.com/smallrye/smallrye-mutiny[Mutiny] is a reactive programming library allowing to express and compose asynchronous actions.
It offers 2 types:

* `io.smallrye.mutiny.Uni` - for asynchronous action providing 0 or 1 result
* `io.smallrye.mutiny.Multi` - for multi-item (with back-pressure) streams

Both types are lazy and follow a subscription pattern.

> The computation only starts once there is an actual need for it (i.e., a subscriber enlists).

Reviewer 2: ```suggestion
The computation only starts once there is an actual need for it (i.e. a subscriber enlists).
```

[source, java]
----
uni.subscribe().with(
    result -> System.out.println("result is " + result),
    failure -> failure.printStackTrace()
);

multi.subscribe().with(
    item -> System.out.println("Got " + item),
    failure -> failure.printStackTrace()
);
----

Both `Uni` and `Multi` expose event-driven APIs: you express what you want to do upon a given event (success, failure, etc.).
These APIs are divided into groups (types of operations) to make it more expressive and avoid having 100s of methods attached to a single class.
The main types of operations are about reacting to failure, completion, manipulating items, extracting, or collecting them.
It provides a smooth coding experience, with a navigable API, and the result does not require too much knowledge around reactive.

[source, java]
----
httpCall
    .onFailure().recoverWithItem("my fallback");
----

You may wonder about Reactive Streams (https://www.reactive-streams.org/).

> `Multi` implements Reactive Streams `Publisher`, and so implements the Reactive Streams back-pressure mechanism.

Reviewer 2: Aren't we going to ditch that at some point?

Author: Publisher is the base, we removed Reactive Streams Operators.

[source, java]
----
Maybe<String> maybe = uni.convert().with(UniRxConverters.toMaybe());
Flux<String> flux = multi.convert().with(MultiReactorConverters.toFlux());
----

But, what about Vert.x?
Vert.x APIs are also available using Mutiny types.
The following snippet shows a usage of the Vert.x Web Client:

[source, java]
----
// Use io.vertx.mutiny.ext.web.client.WebClient
client = WebClient.create(vertx,
                new WebClientOptions().setDefaultHost("fruityvice.com").setDefaultPort(443).setSsl(true)
                        .setTrustAll(true));
// ...
Uni<JsonObject> uni =
    client.get("/api/fruit/" + name)
        .send()
        .onItem().apply(resp -> {
            if (resp.statusCode() == 200) {
                return resp.bodyAsJsonObject();
            } else {
                return new JsonObject()
                        .put("code", resp.statusCode())
                        .put("message", resp.bodyAsString());
            }
        });
----

Last but not least, Mutiny has built-in integration with MicroProfile Context Propagation so you can propagate transactions, traceability data, and so on in your reactive pipeline.

> But enough talking, start to get your hands dirty!

Reviewer 2: ```suggestion
But enough talking, let's get our hands dirty!
```
?

== Bootstrapping the project

The easiest way to create a new Quarkus project is to open a terminal and run the following command:

> For Linux & macOS users

Reviewer 2: ```suggestion
For Linux and macOS users
```
(we should avoid `&` per doc rules)

cd getting-started-reactive
----

For Windows users

- If using cmd, (don't use forward slash `\`)

[source,shell,subs=attributes+]
----
mvn io.quarkus:quarkus-maven-plugin:{quarkus-version}:create -DprojectGroupId=org.acme -DprojectArtifactId=getting-started-reactive -DclassName="org.acme.quickstart.ReactiveGreetingResource" -Dpath="/hello" -Dextensions="resteasy-mutiny,resteasy-jsonb"
----

- If using Powershell, wrap `-D` parameters in double quotes

[source,shell,subs=attributes+]
----
mvn io.quarkus:quarkus-maven-plugin:{quarkus-version}:create "-DprojectGroupId=org.acme" "-DprojectArtifactId=getting-started-reactive" "-DclassName=org.acme.quickstart.ReactiveGreetingResource" "-Dpath=/hello" "-Dextensions=resteasy-mutiny,resteasy-jsonb"
----

It generates the following in  `./getting-started-reactive`:

* the Maven structure
* an `org.acme.quickstart.ReactiveGreetingResource` resource exposed on `/hello`
* an associated unit test
* a landing page that is accessible on `http://localhost:8080` after starting the application
* example `Dockerfile` files for both `native` and `jvm` modes in `src/main/docker`
* the application configuration file

> The generated `pom.xml` also declares the RESTEasy Mutiny support and RESTEasy JSON-B to serialized payloads.

Reviewer 2: ```suggestion
The generated `pom.xml` also declares the RESTEasy Mutiny support and RESTEasy JSON-B to serialize payloads.
```

=== Reactive JAX-RS resources

During the project creation, the `src/main/java/org/acme/quickstart/ReactiveGreetingResource.java` file has been created with the following content:

[source,java]
----
package org.acme.quickstart;

import javax.ws.rs.GET;
import javax.ws.rs.Path;
import javax.ws.rs.Produces;
import javax.ws.rs.core.MediaType;

@Path("/hello")
public class ReactiveGreetingResource {

    @GET
    @Produces(MediaType.TEXT_PLAIN)
    public String hello() {
        return "hello";
    }
}
----

It's a very simple REST endpoint, returning "hello" to requests on "/hello".

> Let's now creates a `ReactiveGreetingService` class with the following content:

Reviewer 2: ```suggestion
Let's now create a `ReactiveGreetingService` class with the following content:
```

import io.smallrye.mutiny.Multi;
import io.smallrye.mutiny.Uni;
import org.jboss.resteasy.annotations.SseElementType;
import org.jboss.resteasy.annotations.jaxrs.PathParam;
import org.reactivestreams.Publisher;

@Path("/hello")
public class ReactiveGreetingResource {

    @Inject
    ReactiveGreetingService service;

    @GET
    @Produces(MediaType.TEXT_PLAIN)
    @Path("/greeting/{name}")
    public Uni<String> greeting(@PathParam String name) {
        return service.greeting(name);
    }

    @GET
    @Produces(MediaType.TEXT_PLAIN)
    public String hello() {
        return "hello";
    }
}
----

The `ReactiveGreetingService` class contains a straightforward method producing a `Uni`.
While, in this example, the resulting item is emitted immediately, you can imagine any async API producing a `Uni`. We cover this later in this guide.

Now, start the application using:

[source, shell]
----
./mvnw quarkus:dev
----

Once running, check that you get the expected greeting message by opening http://localhost:8080/hello/greeting/neo.

== Handling streams
So far, we only return an asynchronous result.
In this section, we extend the application with streams conveying multiple items.

> These streams could come from Kafka and any other source of data, but to keep things simple, we just generate periodic greeting messages.

Reviewer 2: ```suggestion
These streams could come from Kafka or any other source of data, but to keep things simple, we just generate periodic greeting messages.
```

In the `ReactiveGreetingService`, add the following method:

[source, java]
----
public Multi<String> greetings(int count, String name) {

>   return Multi.createFrom().ticks().every(Duration.ofSeconds(1))

Reviewer 2: It's minor but the indentation is off.

Author: Fixed.

@Path("/stream/{count}/{name}")
public Multi<String> greetingsAsStream(@PathParam int count, @PathParam String name) {
    return service.greetings(count, name);
}
----

The only difference with the previous snippet is the produced type and the `@SseElementType` annotation indicating the type of each event.

You can see the result using:

[source, shell]
----
$ curl -N http://localhost:8080/hello/stream/5/neo
data: hello neo - 0

data: hello neo - 1

data: hello neo - 2

data: hello neo - 3

data: hello neo - 4
----

== Using Reactive APIs

=== Using Quarkus reactive APIs

Quarkus provides many reactive APIs using the Mutiny model.
In this section, we are going to see how you can use the Reactive PostgreSQL driver to interact with your database in a non-blocking and reactive way.

Create a new project using:

[source,shell,subs=attributes+]
----
mvn io.quarkus:quarkus-maven-plugin:{quarkus-version}:create \
    -DprojectGroupId=org.acme \
    -DprojectArtifactId=getting-started-reactive-crud \
    -DclassName="org.acme.reactive.crud.FruitResource" \
    -Dpath="/fruits" \
    -Dextensions="resteasy-mutiny, resteasy-jsonb, reactive-pg-client"
cd getting-started-reactive-crud
----

> This application is interacting with a Postgresql database, so you need one:

Reviewer 2: ```suggestion
This application is interacting with a PostgreSQL database, so you need one:
```

[source, shell]
----
docker run --ulimit memlock=-1:-1 -it --rm=true --memory-swappiness=0 \
           --name postgres-quarkus-reactive -e POSTGRES_USER=quarkus_test \
           -e POSTGRES_PASSWORD=quarkus_test -e POSTGRES_DB=quarkus_test \
           -p 5432:5432 postgres:10.5
----

> Then, let's configure our data source.

Reviewer 2: ```suggestion
Then, let's configure our datasource.
```

Open the `src/main/resources/application.properties` and add the following content:

[source, properties]
----
quarkus.datasource.url=vertx-reactive:postgresql://localhost:5432/quarkus_test
quarkus.datasource.username=quarkus_test

> quarkus.datasource.password=quarkus_test

Reviewer 2: ```suggestion
quarkus.datasource.db-kind=postgresql
quarkus.datasource.username=quarkus_test
quarkus.datasource.password=quarkus_test
quarkus.datasource.reactive.url=vertx-reactive:postgresql://localhost:5432/quarkus_test
```
would be the new way of doing things now.

Reviewer 2: IIRC, you can even remove the `vertx-reactive:` part when doing it this way.

Author: I will try tomorrow morning.

Author: It's working, thanks!

myapp.schema.create=true
----

> The 3 first lines define the data source.

Reviewer 2: ```suggestion
The 3 first lines define the datasource.
```

> Note database URL prefix, using `vertx-reactive` instead of the traditional `jdbc`.

Reviewer 2: ```suggestion
Note the database URL prefix, using `vertx-reactive` instead of the traditional `jdbc`.
```

Reviewer 2: If you get rid of the prefix, you can remove this note.

        return client.preparedQuery("SELECT id, name FROM fruits WHERE id = $1", Tuple.of(id))
                .onItem().apply(RowSet::iterator)
                .onItem().apply(iterator -> iterator.hasNext() ? from(iterator.next()) : null);
    }

    public Uni<Long> save(PgPool client) {
        return client.preparedQuery("INSERT INTO fruits (name) VALUES ($1) RETURNING (id)", Tuple.of(name))
                .onItem().apply(pgRowSet -> pgRowSet.iterator().next().getLong("id"));
    }

    public Uni<Boolean> update(PgPool client) {
        return client.preparedQuery("UPDATE fruits SET name = $1 WHERE id = $2", Tuple.of(name, id))
                .onItem().apply(pgRowSet -> pgRowSet.rowCount() == 1);
    }

    public static Uni<Boolean> delete(PgPool client, Long id) {
        return client.preparedQuery("DELETE FROM fruits WHERE id = $1", Tuple.of(id))
                .onItem().apply(pgRowSet -> pgRowSet.rowCount() == 1);
    }

    private static Fruit from(Row row) {
        return new Fruit(row.getLong("id"), row.getString("name"));
    }
}
----

This _entity_ contains a few fields and methods to find, update, and delete rows from the database.

> These methods return either `Unis` or `Multis` as the produced items are emitted asynchronously when the database has retrieved the result.

Reviewer 2: this is a bit weird `when the database has retrieved the result`. Is the database retrieving things?

Author: changing it to ` when the results have been retrieved.`

Notice that the reactive PostgreSQL client already provides `Uni` and `Multi` instances.
So you only transform the results from the database into _business-friendly_ objects.

Then, let's use this `Fruit` class in the `FruitResource`.
Edit the `FruitResource` class to match the following content:

[source, java]
----
package org.acme.reactive.crud;

import io.smallrye.mutiny.Multi;
import io.smallrye.mutiny.Uni;
import io.vertx.mutiny.pgclient.PgPool;
import org.eclipse.microprofile.config.inject.ConfigProperty;
import org.jboss.resteasy.annotations.jaxrs.PathParam;

import javax.annotation.PostConstruct;
import javax.inject.Inject;

> import javax.ws.rs.*;

Reviewer 2: We should avoid star imports.

Author: Yes, got bitten by the check on the quickstart.

|===
|groupId:artifactId | Description

|`io.smallrye.reactive:smallrye-mutiny-vertx-core`
|Mutiny API for Vert.x Core

|`io.smallrye.reactive:smallrye-mutiny-vertx-mail-client`
|Mutiny API for the Vert.x Mail Client

|`io.smallrye.reactive:smallrye-mutiny-vertx-web-client`
|Mutiny API for the Vert.x Web Client

|`io.smallrye.reactive:smallrye-mutiny-vertx-mongo-client`
|Mutiny API for the Vert.x Mongo Client

|`io.smallrye.reactive:smallrye-mutiny-vertx-redis-client`
|Mutiny API for the Vert.x Redis Client

|`io.smallrye.reactive:smallrye-mutiny-vertx-cassandra-client`
|Mutiny API for the Vert.x Cassandra Client

|`io.smallrye.reactive:smallrye-mutiny-vertx-consul-client`
|Mutiny API for the Vert.x Consul Client

|`io.smallrye.reactive:smallrye-mutiny-vertx-kafka-client`
|Mutiny API for the Vert.x Kafka Client

|`io.smallrye.reactive:smallrye-mutiny-vertx-amqp-client`
|Mutiny API for the Vert.x AMQP Client

|`io.smallrye.reactive:smallrye-mutiny-vertx-rabbitmq-client`
|Mutiny API for the Vert.x RabbitMQ Client

|`io.smallrye.reactive:smallrye-mutiny-vertx-rabbitmq-client`
|Mutiny API for the Vert.x RabbitMQ Client

|`io.smallrye.reactive:smallrye-mutiny-vertx-rabbitmq-client`

> |Mutiny API for the Vert.x RabbitMQ Client

Reviewer 2: You have the RabbitMQ work three times.

Author: You know I love rabbits!

                        .setTrustAll(true));
    }

    @GET
    @Produces(MediaType.APPLICATION_JSON)
    @Path("/{name}")
    public Uni<JsonObject> getFruitData(@PathParam("name") String name) {
        return client.get("/api/fruit/" + name)
                .send()
                .map(resp -> {
                    if (resp.statusCode() == 200) {
                        return resp.bodyAsJsonObject();
                    } else {
                        return new JsonObject()
                                .put("code", resp.statusCode())
                                .put("message", resp.bodyAsString());
                    }
                });
    }

}
----

There are 2 important points:

1. The injected Vert.x instance has the `io.vertx.mutiny.core.Vertx` type which is the Mutiny variant of Vert.x;
2. The Web Client is created from `io.vertx.mutiny.ext.web.client.WebClient`.

The Mutiny version of the Vert.x APIs also offers:

* `andAwait` methods such as `sendAndAwait`. `andAwait` indicates that the caller thread is blocked until the result is available.
Be aware not to block the event loop / IO thread that way.

> * `andForget` methods such as `writeAndForget`. `andForget` is available for method returning an `Uni`.

Reviewer 2: ```suggestion
* `andForget` methods such as `writeAndForget`. `andForget` is available for method returning a `Uni`.
```

`andForget` indicates that you don't need the resulting `Uni` indicating the success or failure of the operation.
However, remember that if you don't subscribe, the operation would not be triggered.
`andForget` manages this for you and manage the subscription.
* `toMulti` methods allowing to transform a Vert.x `ReadStream` into a `Multi`
* `toBlockingIterable` / `toBlockingStream` methods allowing to transform a Vert.x `ReadStream` into a blocking iterable or blocking `java.util.Stream`

=== Using RxJava or Reactor APIs

> Mutiny provides utilities to convert Rx Java 2 and Project Reactor types to `Uni` and `Multi`.

Reviewer 2: ```suggestion
Mutiny provides utilities to convert RxJava 2 and Project Reactor types to `Uni` and `Multi`.
```

> Rx Java 2 converters are available in the following dependency:

Reviewer 2: ```suggestion
RxJava 2 converters are available in the following dependency:
```

[source, xml]
----
<dependency>
    <groupId>io.smallrye.reactive</groupId>
    <artifactId>mutiny-rxjava</artifactId>
    <version>$VERSION</version>
</dependency>
----

> So if you have an API returning RxJava 2 types (`Completable`, `Single`, `Maybe`, `Observable`, `Flowable`), you can creates `Unis` and `Multis` as follows:

Reviewer 2: ```suggestion
So if you have an API returning RxJava 2 types (`Completable`, `Single`, `Maybe`, `Observable`, `Flowable`), you can create `Unis` and `Multis` as follows:
```

Uni<String> uniFromSingle = Uni.createFrom().converter(UniRxConverters.fromSingle(), single);
Uni<String> uniFromMaybe = Uni.createFrom().converter(UniRxConverters.fromMaybe(), maybe);
Uni<String> uniFromEmptyMaybe = Uni.createFrom().converter(UniRxConverters.fromMaybe(), emptyMaybe);
Uni<String> uniFromObservable = Uni.createFrom().converter(UniRxConverters.fromObservable(), observable);
Uni<String> uniFromFlowable = Uni.createFrom().converter(UniRxConverters.fromFlowable(), flowable);

Multi<Void> multiFromCompletable = Multi.createFrom().converter(MultiRxConverters.fromCompletable(), completable);
Multi<String> multiFromSingle = Multi.createFrom().converter(MultiRxConverters.fromSingle(), single);
Multi<String> multiFromMaybe = Multi.createFrom().converter(MultiRxConverters.fromMaybe(), maybe);
Multi<String> multiFromEmptyMaybe = Multi.createFrom().converter(MultiRxConverters.fromMaybe(), emptyMaybe);
Multi<String> multiFromObservable = Multi.createFrom().converter(MultiRxConverters.fromObservable(), observable);
Multi<String> multiFromFlowable = Multi.createFrom().converter(MultiRxConverters.fromFlowable(), flowable);
----

> You can also transform `Unis` and `Multis` into Rx Java types:

Reviewer 2: ```suggestion
You can also transform `Unis` and `Multis` into RxJava types:
```

Single<Optional<String>> single = uni.convert().with(UniRxConverters.toSingle());
Single<String> single2 = uni.convert().with(UniRxConverters.toSingle().failOnNull());
Maybe<String> maybe = uni.convert().with(UniRxConverters.toMaybe());
Observable<String> observable = uni.convert().with(UniRxConverters.toObservable());
Flowable<String> flowable = uni.convert().with(UniRxConverters.toFlowable());
// ...
Completable completable = multi.convert().with(MultiRxConverters.toCompletable());
Single<Optional<String>> single = multi.convert().with(MultiRxConverters.toSingle());
Single<String> single2 = multi.convert().with(MultiRxConverters
        .toSingle().onEmptyThrow(() -> new Exception("D'oh!")));
Maybe<String> maybe = multi.convert().with(MultiRxConverters.toMaybe());
Observable<String> observable = multi.convert().with(MultiRxConverters.toObservable());
Flowable<String> flowable = multi.convert().with(MultiRxConverters.toFlowable());
----

Project Reactor converters are available in the following dependency:

[source, xml]
----
<dependency>
    <groupId>io.smallrye.reactive</groupId>
    <artifactId>mutiny-reactor</artifactId>
    <version>$VERSION</version>
</dependency>
----

> So if you have an API returningReactor types (`Mono`, `Flux`), you can creates `Unis` and `Multis` as follows:

Reviewer 2: ```suggestion
So if you have an API returning Reactor types (`Mono`, `Flux`), you can create `Unis` and `Multis` as follows:
```
